### PR TITLE
Only retrieve study modes of course options with vacancies

### PR DESCRIPTION
**Pull request by despo**
Resolves sentry error caused by candidate ending up on location selection page with no locations rendered due to the selected study mode not having any sites with availability.

https://sentry.io/organizations/dfe-bat/issues/2297226890/?project=1765973&referrer=slack

A new method is introduced as the existing `available_study_mode_from_options` is used on Make Offer, on the provider side, where availability should not be taken into account.

### CODE REVIEW

File: spec/models/course_spec.rb [hunk 1]
```diff
@@ -47,6 +45,17 @@
     end
   end
 
+  describe '#available_study_modes_with_vacancies' do
+    let(:course) { build(:course) }
+
+    it 'returns an array of unique study modes for course options with available vacancies' do
```
Comment (cwrw) [inline spec/models/course_spec.rb:51]: can you add a spec for the `site_still_valid` option as well?

**Review by cwrw — APPROVED** [review]
Looks good 👍  Might be worth getting someone from the candidate side as well to confirm

**Review by stevehook — COMMENTED** [review]
I think this change makes sense.

I think I found another variation on the original bug though when testing locally. I've reproduced on the review app by changing the availability of _Blackfriars Teaching School Alliance (1JA)_ part-time course option for _English and Communication (SEND) (V252)_ to `no_vacancies` and try to add that course. I end up getting redirected to the empty site selection page.

image

Pressing continue triggers a 500.

I think part of the problem here is that we are in a 'single-site' situation here but https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/controllers/candidate_interface/course_choices/course_selection_controller.rb#L50 doesn't work because there are multiple options (due to different study modes).

**Comment by despo** [comment]
@stevehook thanks for pointing this out. Looking more into it I think all course option retrievals in the `pick_*` forms should be updated to ensure that they satisfy the conditions specified in `available_study_modes_with_vacancies`.

File: app/models/course.rb [hunk 2]
```diff
@@ -96,6 +96,10 @@ def available_study_modes_from_options
     course_options.select(&:site_still_valid).pluck(:study_mode).uniq
   end
 
+  def available_study_modes_with_vacancies
+    course_options.available.selectable.pluck(:study_mode).uniq
```
Comment (malcolmbaig) [inline app/models/course.rb:100]: We can just do `course_options.available`, as the `available` scope already uses `selectable`.

Comment (despo) [reply app/models/course.rb:100]: Thanks for pointing out. Will have a look in a bit.

**Comment by malcolmbaig** [comment]
> @stevehook thanks for pointing this out. Looking more into it I think all course option retrievals in the `pick_*` forms should be updated to ensure that they satisfy the conditions specified in `available_study_modes_with_vacancies`.

💯 — I think there are two spots where we need additional `available` scopes:

https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/forms/candidate_interface/pick_course_form.rb#L48
https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/forms/candidate_interface/pick_study_mode_form.rb#L9

**Comment by despo** [comment]
@stevehook retested with the latest updates using the details you provided and things seem to work now.

**Review by stevehook — APPROVED** [review]
Re-tested review app and all works now. Nice fix! 💯

**Pull request merged**